ksql-plugins/ksql-rocksdb-config-setter/src/main/java/io/confluent/ksql/rocksdb/KsqlBoundedMemoryRocksDBConfig.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rocksdb;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.kafka.streams.state.RocksDBConfigSetter;
+import org.rocksdb.BlockBasedTableConfig;
+import org.rocksdb.Options;
+
+/**
+ * This {@code RocksDBConfigSetter} implementation limits the total memory used
+ * across all RocksDB instances to the number of bytes passed via
+ * "ksql.plugins.rocksdb.total.memory", and also configures
+ * the shared RocksDB thread pool to use "ksql.plugins.rocksdb.num.background.threads" threads.
+ * </p>
+ * See https://docs.confluent.io/5.3.0/streams/developer-guide/memory-mgmt.html#rocksdb.
+ */
+public class KsqlBoundedMemoryRocksDBConfig implements RocksDBConfigSetter {
+
+  private static final String CONFIG_PREFIX = "ksql.plugins.rocksdb.";
+  private static final String TOTAL_OFF_HEAP_MEMORY_CONFIG = CONFIG_PREFIX + "total.memory";
+  private static final String N_BACKGROUND_THREADS_CONFIG =
+      CONFIG_PREFIX + "num.background.threads";
+
+  private static final double INDEX_FILTER_BLOCK_RATIO = 0.1;
+  private static final int N_MEMTABLES = 6;
+  private static final long BLOCK_SIZE = 4096L;
+
+  private static org.rocksdb.Cache cache;
+  private static org.rocksdb.WriteBufferManager writeBufferManager;
+
+  private static AtomicBoolean configured = new AtomicBoolean(false);
+  private static long totalOffHeapMemory;
+  private static long memtableSize;
+  private static long totalMemtableMemory;
+  private static int numBackgroundThreads;
+
+  public static void configure(final Map<String, Object> config) {
+    configure(config, new Options());
+  }
+
+  @VisibleForTesting
+  static void configure(final Map<String, Object> config, final Options options) {
+    if (configured.getAndSet(true)) {
```

Comment:
There's an edge case here where if `setConfig()` is called after `configure()` is started but before it has finished, we may get undefined behavior. This shouldn't happen since `configure()` is called as part of the server starting up while `setConfig()` is only called when running streams apps, but if anyone has suggestions for improving the checks in these files to handle this edge case I'm all ears!